File: bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/BazelInfoResolver.java
package org.jetbrains.bsp.bazel.bazelrunner;

import io.vavr.Lazy;
import io.vavr.Tuple2;
import io.vavr.collection.Map;
import io.vavr.control.Option;
import java.nio.file.Paths;
import java.util.function.Function;
import java.util.regex.Pattern;

public class BazelInfoResolver {
  private static final Pattern INFO_LINE_PATTERN = Pattern.compile("([\\w-]+): (.*)");
  private final BazelRunner bazelRunner;
  private final BazelInfoStorage storage;

  public BazelInfoResolver(BazelRunner bazelRunner, BazelInfoStorage storage) {
    this.bazelRunner = bazelRunner;
    this.storage = storage;
  }

  public BazelInfo resolveBazelInfo() {
    return new LazyBazelInfo(Lazy.of(() -> storage.load().getOrElse(this::bazelInfoFromBazel)));
  }

  private BazelInfo bazelInfoFromBazel() {
    var processResult =
        bazelRunner.commandBuilder().info().executeBazelCommand().waitAndGetResult();
    var info = parseBazelInfo(processResult);
    storage.store(info);
    return info;
  }

  private BasicBazelInfo parseBazelInfo(BazelProcessResult bazelProcessResult) {
    var outputMap =
        bazelProcessResult
            .stdoutLines()
            .flatMap(
                line -> {
                  var matcher = INFO_LINE_PATTERN.matcher(line);
                  return Option.when(
                      matcher.matches(), () -> new Tuple2<>(matcher.group(1), matcher.group(2)));
                })
            .toMap(Function.identity());

    var executionRoot = extract(outputMap, "execution_root");
    var workspace = Paths.get(extract(outputMap, "workspace"));
    return new BasicBazelInfo(executionRoot, workspace);
  }

  private String extract(Map<String, String> outputMap, String name) {
    return outputMap
        .get(name)
        .getOrElseThrow(
            () ->
                new RuntimeException(String.format("Failed to resolve %s from bazel info", name)));
  }
}

File: bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/BazelInfoResolver.kt
package org.jetbrains.bsp.bazel.bazelrunner

import org.eclipse.lsp4j.jsonrpc.CancelChecker
import org.jetbrains.bsp.bazel.commons.escapeNewLines
import java.nio.file.Paths

class BazelInfoResolver(
    private val bazelRunner: BazelRunner,
    private val storage: BazelInfoStorage
) {

  fun resolveBazelInfo(cancelChecker: CancelChecker): BazelInfo {
    return LazyBazelInfo { storage.load() ?: bazelInfoFromBazel(cancelChecker) }
  }

  private fun bazelInfoFromBazel(cancelChecker: CancelChecker): BazelInfo {
    val processResult = bazelRunner
          .commandBuilder()
          .info()
          .executeBazelCommand()
          .waitAndGetResult(cancelChecker,true)
    return parseBazelInfo(processResult).also { storage.store(it) }
  }

  private fun parseBazelInfo(bazelProcessResult: BazelProcessResult): BasicBazelInfo {
    val outputMap = bazelProcessResult
        .stdoutLines
        .mapNotNull { line ->
          InfoLinePattern.matchEntire(line)?.let { it.groupValues[1] to it.groupValues[2] }
        }.toMap()

    fun extract(name: String): String =
        outputMap[name]
            ?: error("Failed to resolve $name from bazel info in ${bazelRunner.workspaceRoot}. " +
                "Bazel Info output: '${bazelProcessResult.stdout.escapeNewLines()}'")

    return BasicBazelInfo(
        execRoot = extract("execution_root"),
        outputBase = Paths.get(extract("output_base")),
        workspaceRoot = Paths.get(extract("workspace")),
        release = BazelRelease.fromReleaseString(extract("release"))
    )
  }

  companion object {
    private val InfoLinePattern = "([\\w-]+): (.*)".toRegex()
  }
}
